### DISCUSSION TITLE
ui.add_head_html <title> has no effect

### DISCUSSION THREAD

**Discussion by xiaozhiagi** [zauberzeug/nicegui#1436] [Ideas / Feature Requests]
### Question

Due to the inability to set the title in advance, I would like to use ui.add_head_html setting title, but it did not work

```python

@ui.page('/phm/{root}/{topic}/{f}', title=None)
def phmai_vibrate(root, topic, f):
      ui.add_head_html(f'''
        <title>{root} | {topic}</title>
        ''')
      pass
```

**Comment by rodja** [comment]
You should use js for that:

```py
    async def set_title() -> None:
        await ui.run_javascript('document.title = "A New Title"')
    ui.button('change title', on_click=set_title)
```

**Reply by xiaozhiagi** [reply]
Although this click trigger can be changed, for more practical and dynamic requirements, it is recommended that future versions of Nicegui consider using ui.add_ Head_ Add to the HTML or dynamically generate a title when clicking on the URL

**Reply by ed2050** [reply]
Setting dynamic title is a commonly needed function for web apps and websites.  This should be a first class feature of any web toolkit.

Suggest converting this thread to a Feature Request.

**Reply by falkoschindler** [reply]
@ed2050 You're right, changing a page title after defining a route (e.g. depending on path parameters) would be very useful. We already have a solution in mind which I will post as a pull request in the next few days.

**Reply by iron3oxide** [reply]
Really need this for my project, great to hear there's a solution soon!

**Reply by ed2050** [reply]
Thanks @falkoschindler, great that you're working on a solution.  Let me illustrate one problem I have to see if the proposed solution covers it.

# Problem
Say I have a search results page.  The generic title is Results.  But that's not useful when you have multiple tabs / bookmarks of search results.  Full title should include the search terms, like "Results : terms ...".  So there's a fixed part and a variable part.

Currently you can set the `<title>` tag in HTML head by passing an argument to ui.page.  So you can do `@ui.page ('/results', title = 'Results')`.  Then update title later inside page function with javascript.

Unfortunately @ui.page does _not_ pass the title to the page function.  Everything else gets passed: path params, query params, nicegui client object, fastapi request object.  But not the title.  This makes it awkward and difficult for the page function to get the fixed part of the title.  Options are:
1. Store fixed portion of title in a global variable where both ui.page and page function can access it
2. Wait for client.connected and use js to read the fixed portion from the DOM title

2 is a mess so let's focus on 1.  Here's sample code:

```
pagetitle = 'Results'

@ui.page ('/results', title = pagetitle)
async def resultspage (client : nicegui.Client, request : Request, terms, title = pagetitle) :
    'page with search results'
    newtitle = f'{ title } : { terms }'
    # wait for client.connected, set document title with js
``` 

Not bad, right?  Just an extra global var for `pagetitle`.  Have to set `title = pagetitle` as default value to resultspage because @ui.page will never pass it.  Or you could access the global pagetitle directly.  Doesn't matter, same issue.

The problem is, it gets messy when you have mutiple page functions in one module.  Now imagine this:
```
pagetitle = 'Some page'
@ui.page ('/results', title = pagetitle)
async def somepage (client : nicegui.Client, request : Request, terms, title = pagetitle) :

pagetitle = 'Other page'
@ui.page ('/results', title = pagetitle)
async def otherpage (client : nicegui.Client, request : Request, terms, title = pagetitle) :

pagetitle = 'Another page'
@ui.page ('/results', title = pagetitle)
async def anotherpage (client : nicegui.Client, request : Request, terms, title = pagetitle) :

@ui.page ('/results', title = pagetitle)
async def resultspage (client : nicegui.Client, request : Request, terms, title = pagetitle) :
```

What a mess.  Either pagetitle gets redefined before every page function.  Or you use different names on every pagefunction: pagetitle1, pagetitle2, pagetitle3, etc.  Both are bad practices: mistake prone with needless repetition.

Did you notice I forgot to redefine pagetitle before resultspage?  Now resultspage will show "Another page" as its title.  Oops!  Such mistakes are common with this approach.

If ui.page just passes the title param to the page function like it already does with other kinds of parameters, then everything becomes much simpler:
```
@ui.page ('/results', title = 'Some page')
async def somepage (client : nicegui.Client, request : Request, terms, title) :

@ui.page ('/results', title = 'Other page')
async def otherpage (client : nicegui.Client, request : Request, terms, title) :

# no more globals!  huzzah!
```

I hope your solution being developed considers this problem where ui.page doesn't communicate the title it takes with the decorated page function.  Thanks.

**Reply by falkoschindler** [reply]
Our plan was to set the `title` on the `client` object. This way the page function has automatically access to it. Unfortunately I haven't yet found the time for the pull request. But it's on my list. 🤞

**Reply by ed2050** [reply]
Great, that should work.  No rush.  Good solutions are worth waiting for. 😄

**Reply by falkoschindler** [reply]
I just noticed that you can already access the page title like this:
```py
@ui.page('/data', title='Data page')
def data_page(client: Client):
    ui.label(f'Title: {client.page.title}')
```

**Comment by ed2050** [comment]
Nice!  I hope to see client better documented so people can know what it
offers. :)

**Comment by falkoschindler** [comment]
Version 1.4.6 with `ui.page_title` has been release. I'll close this feature request.
